src/main/java/emu/grasscutter/game/quest/GameQuest.java

```diff
@@ -199,6 +199,13 @@ public void fail() {
 
         getQuestData().getFailExec().forEach(e -> getOwner().getServer().getQuestSystem().triggerExec(this, e, e.getParam()));
 
+        // TODO, QUEST_EXEC_REMOVE_TRIAL_AVATAR only triggers when quest is finished, 
+        // so the following code must also be placed somewhere on quest fails,
+        // not sure if this is the right placed though
+        if (getQuestData().getTrialAvatarList() != null) {
```

Comment:
I don't think this should be handled manually, most main quests have multiple ways to trigger `QUEST_EXEC_REMOVE_TRIAL_AVATAR`, either via a failExec, or via leave dungeon quest, that also rewinds the quest.